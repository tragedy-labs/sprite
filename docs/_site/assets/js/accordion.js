/*
 *   This content is licensed according to the W3C Software License at
 *   https://www.w3.org/copyright/software-license-2023/
 *
 *   Simple accordion pattern example
 */

'use strict';

class Accordion {
  constructor(domNode) {
    this.root = domNode;
    this.button = this.root.querySelector('button[aria-expanded]');
    this.icon = this.root.querySelector('svg');
    const controlsId = this.button.getAttribute('aria-controls');
    this.content = document.getElementById(controlsId);

    this.open = this.button.getAttribute('aria-expanded') === 'true';

    // add event listeners
    this.button.addEventListener('click', this.onButtonClick.bind(this));
    this.toggle(false);
  }

  onButtonClick() {
    this.toggle(!this.open);
  }

  toggle(open) {
    // don't do anything if the open state doesn't change
    if (open === this.open) {
      return;
    }

    // update the internal state
    this.open = open;

    // handle DOM updates
    this.button.setAttribute('aria-expanded', `${open}`);
    if (open) {
      this.content.removeAttribute('hidden');
      this.icon.classList.add('open');
    } else {
      this.content.setAttribute('hidden', '');
      this.icon.classList.remove('open');
    }
  }

  // Add public open and close methods for convenience
  open() {
    this.toggle(true);
  }

  close() {
    this.toggle(false);
  }
}

const initializeAccordions = () => {
  // init accordions
  const accordions = document.querySelectorAll('.menuItem');
  accordions.forEach((accordionEl) => {
    new Accordion(accordionEl);
  });
};

document.addEventListener('DOMContentLoaded', initializeAccordions);
