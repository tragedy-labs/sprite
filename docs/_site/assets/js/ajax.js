// Constants
const CONTENT_CONTAINER_ID = "content";
const PARSER = new DOMParser();

// State
let historyStack = [];
let currentStateIndex = -1;

// Functions
async function fetchDocument(url) {
  try {
    const response = await fetch(url);
    return await response.text();
  } catch (error) {
    throw new Error(`Failed to fetch document: ${url}`, { cause: error });
  }
}

async function parseHtml(html) {
  return PARSER.parseFromString(html, "text/html");
}

async function updateContent(htmlDoc) {
  const title = htmlDoc.querySelector("title").textContent;
  const contentDiv = htmlDoc.getElementById(CONTENT_CONTAINER_ID);
  const target = document.getElementById(CONTENT_CONTAINER_ID);
  target.innerHTML = contentDiv.innerHTML;
  target.scrollTop = 0;
  document.title = title;

  // Reinitialize link click events for newly loaded content
  createLinkClickEvents(target);
}

function pushToHistory(url) {
  historyStack = historyStack.slice(0, currentStateIndex + 1);
  historyStack.push(url);
  currentStateIndex++;
  history.pushState({ index: currentStateIndex }, "", url);
}

async function navigate(url, addToHistory = true) {
  const urlObj = new URL(url, window.location.origin);

  if (urlObj)
    if (urlObj.hash && urlObj.pathname === window.location.pathname) {
      // Handle local page anchor navigation
      const anchor = document.querySelector(urlObj.hash);
      if (anchor) {
        anchor.scrollIntoView();
        addToHistory && pushToHistory(url);
        return;
      }
    }

  // Handle Local Navigation
  const html = await fetchDocument(url);
  const htmlDoc = await parseHtml(html);
  await updateContent(htmlDoc);

  addToHistory && pushToHistory(url);
}

function createLinkClickEvents(root) {
  const links = root.querySelectorAll("a");
  links.forEach((link) => {
    link.addEventListener("click", (e) => {
      const url = e.currentTarget.href;
      if (isLocalLink(url)) {
        e.preventDefault();
        navigate(url);
      }
    });
  });
}

function isLocalLink(url) {
  const linkUrl = new URL(url, window.location.origin);
  return linkUrl.origin === window.location.origin;
}

window.addEventListener("popstate", (e) => {
  if (e.state && e.state.index !== undefined) {
    currentStateIndex = e.state.index;
    navigate(historyStack[currentStateIndex], false);
  }
});

// Initialize
function initAjax() {
  createLinkClickEvents(document);
  // Initialize the first state
  historyStack.push(window.location.href);
  currentStateIndex++;
  history.replaceState({ index: currentStateIndex }, "", window.location.href);
}

// Call init
document.addEventListener("DOMContentLoaded", initAjax);
