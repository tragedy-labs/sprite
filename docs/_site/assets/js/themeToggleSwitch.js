function initializeThemeToggle() {
  const element = document.getElementById("theme_toggle");
  const theme = localStorage.getItem("theme");

  const prefersTheme = (theme) =>
    window.matchMedia(`(prefers-color-scheme: ${theme})`).matches;

  const setTheme = (theme) => {
    localStorage.setItem("theme", theme);
    document.body.className = theme;
    setSwitchState(theme);
  };

  const setSwitchState = (theme) => {
    switch (theme) {
      case "dark":
        element.checked = true;
        break;
      case "light":
        element.checked = false;
        break;
      default:
        throw new Error("unknown theme");
    }
  };

  const determinePosition = () => {
    if (theme) {
      setTheme(theme);
    } else {
      if (prefersTheme("dark")) {
        setTheme("dark");
        return;
      } else {
        setTheme("light");
      }
    }
  };

  element.addEventListener("click", () => {
    if (element.checked) {
      setTheme("dark");
    } else {
      setTheme("light");
    }
  });

  determinePosition();
}

document.addEventListener("DOMContentLoaded", initializeThemeToggle);
