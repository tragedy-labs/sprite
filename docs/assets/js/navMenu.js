document.addEventListener("DOMContentLoaded", function () {
  const content = document.getElementById("content");
  const mainNav = document.getElementById("main_nav");
  const navButton = document.getElementById("nav_button");
  const navIndicator = navButton.querySelector("svg");
  const links = mainNav.querySelectorAll("a");

  const closeIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="21"></line>
    <line x1="21" y1="4" x2="3" y2="21"></line>
  `;

  const openIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="4"></line>
    <line x1="3" y1="12" x2="21" y2="12"></line>
    <line x1="3" y1="20" x2="21" y2="20"></line>
  `;

  let isOpen = false;
  let currentTranslation = 0;
  let previousTouchPosition = null;

  function handleTouchEnd() {
    previousTouchPosition = null;

    if (currentTranslation < -80) {
      closeMenu();
    }
    currentTranslation = 0;
    mainNav.style.transform = null;
  }

  function handleTouchStart(event) {
    previousTouchPosition = event.touches[0].clientX;
  }

  function handleTouchMove(event) {
    const currentTouchPosition = event.touches[0].clientX;
    const delta = currentTouchPosition - previousTouchPosition;
    previousTouchPosition = currentTouchPosition;

    currentTranslation += delta;

    if (currentTranslation > 0) {
      currentTranslation = 0;
    }

    mainNav.style.transform = `translate3d(${currentTranslation}px, 0, 0)`;

    const mainNavWidth = mainNav.offsetWidth;
    const ratio = (mainNavWidth + currentTranslation) / mainNavWidth;

    content.style.opacity = 0.65 * (1 - ratio) + 0.35;
  }

  function escHandler(e) {
    if (e.key === "Escape") {
      closeMenu();
    }
  }

  function enableTouchEvents() {
    mainNav.addEventListener("touchstart", handleTouchStart);
    mainNav.addEventListener("touchmove", handleTouchMove);
    mainNav.addEventListener("touchend", handleTouchEnd);
    content.addEventListener("touchstart", closeMenu);
    content.addEventListener("mousedown", closeMenu);
    document.addEventListener("keydown", escHandler);
    // TODO: This could be handled in the ajax script
    // since it already handles link clicks
    links.forEach((link) => {
      link.addEventListener("click", closeMenu);
    });
  }

  function disableTouchEvents() {
    mainNav.removeEventListener("touchstart", handleTouchStart);
    mainNav.removeEventListener("touchmove", handleTouchMove);
    mainNav.removeEventListener("touchend", handleTouchEnd);
    content.removeEventListener("touchstart", closeMenu);
    content.removeEventListener("mousedown", closeMenu);
    document.removeEventListener("keydown", escHandler);
    links.forEach((link) => {
      link.removeEventListener("click", closeMenu);
    });
  }

  function closeMenu() {
    toggleNavState();
    disableTouchEvents();
    navIndicator.innerHTML = openIconMarkup;
    content.style.opacity = null;
  }

  function openMenu() {
    toggleNavState();
    enableTouchEvents();
    navIndicator.innerHTML = closeIconMarkup;
  }

  function toggleNavState() {
    content.classList.toggle("fade");
    mainNav.classList.toggle("open");
    isOpen = !isOpen;
  }

  function handleNavButtonClick() {
    if (isOpen) {
      closeMenu();
    } else {
      openMenu();
    }
  }

  navButton.addEventListener("click", handleNavButtonClick);
});
