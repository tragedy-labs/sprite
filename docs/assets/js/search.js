import data from "/assets/js/searchData.js";
import serializedIndex from "./assets/js/serialized.js";
//import lunr from "lunr"; // assuming lunr is a separate module

// Constants
const MIN_SEARCH_LENGTH = 3;

// State definition
const state = {
  filters: {
    class: true,
    method: true,
    type: true,
  },
  results: {
    filtered: [],
    full: [],
  },
  visible: false,
};

// Elements for reference
const elements = {
  input: null,
  resultsBox: null,
  resultList: null,
  numberOfResults: null,
  numberResultsShowing: null,
  backdrop: null,
  filters: null,
  searchLabel: null,
  close: null,
};

// Initialize search
async function initializeSearch() {
  const root = document.getElementById("search_button");
  elements.input = document.getElementById("searchInput");
  elements.resultsBox = document.getElementById("searchResults");
  elements.resultList = document.getElementById("resultList");
  elements.numberOfResults = document.getElementById("numberOfResults");
  elements.numberResultsShowing = document.getElementById("numberResultsShowing");
  elements.filters = document.getElementById("searchFilters")
    .getElementsByTagName("input");
  elements.searchLabel = document.getElementById("searchLabel");
  elements.close = elements.resultsBox.querySelector("#close_search_dialog");

  // Add event listeners
  elements.input.addEventListener("input", search);
  elements.close.addEventListener("click", hideSearchBox);

  const openSearch = document.getElementById('search_button');

  openSearch.addEventListener('click', showResultsBox);
  document.addEventListener('keydown', (e)=>{
    if (e.key === '/') {
      e.preventDefault();
      showResultsBox();
    };
  }, false);

  // Load the index
  const index = lunr.Index.load(serializedIndex);

  // Initialize search filter toggles
  for (let i = 0; i < elements.filters.length; ++i) {
    elements.filters[i].addEventListener("change", () => {
      state.filters[elements.filters[i].value] = elements.filters[i].checked;
      updateResultList();
    });
  }

  // Define functions
  function appendBackdrop() {
    elements.backdrop = document.createElement("div");
    elements.backdrop.id = "backdrop";
    elements.backdrop.addEventListener("click", hideSearchBox);
    document.body.appendChild(elements.backdrop);
  }

  function removeBackdrop() {
    elements.backdrop.removeEventListener("click", appendBackdrop);
    elements.backdrop.remove();
    elements.backdrop = undefined;
  }

  function handleKeyboardSelection() {
    const items = elements.list.querySelectorAll("a");
    if (items.length > 0) {
      if (document.activeElement === elements.input) {
        e.preventDefault();
        items[0].focus();
      } else {
        for (let i = 0; i < results.length; i++) {
          if (items[i] === document.activeElement) {
            e.preventDefault();
            if (i + 1 < items.length) {
              items[i + 1].focus();
            } else {
              elements.input.focus();
            }
            break;
          }
        }
      }
    }
  }

  function checkKeyPress(e) {
    if (e.key === "Tab") {
      handleKeyboardSelection(e);
    }
    if (e.key === "Escape") {
      hideSearchBox();
    }
  }

  function showResultsBox() {
    appendBackdrop();
    state.visible = true;
    elements.resultsBox.classList.remove("hidden");
    elements.input.focus();
    document.addEventListener("keydown", checkKeyPress);
  }

  function hideSearchBox() {
    state.visible = false;
    elements.resultsBox.classList.add("hidden");
    document.removeEventListener("keydown", checkKeyPress);
    removeBackdrop();
  }

  function updateResultList() {
    state.results.filtered = [];
    elements.resultList.innerHTML = "";
    // elements.numberOfResults.innerHTML = state.results.full.length;
    // elements.numberResultsShowing.innerHTML = state.results.filtered.length;

    if (state.results.full.length > 0) {
      state.results.full.forEach((result) => {
        const doc = data.find((doc) => doc.name === result.ref);
        const type = doc.type === "interface" ? "type" : doc.type;
        if (state.filters[type]) {
          state.results.filtered.push(result.ref);
          const section = document.createElement("section");
          const desc = document.createElement("p");
          const link = document.createElement("a");
          const li = document.createElement("li");
          li.classList.add(doc.type);

          link.href = `/sprite/doc/${doc.name}`;
          link.className = "searchResultLink";
          link.addEventListener("click", () => {});
          link.textContent = doc.parent
            ? `${doc.parent}.${doc.name}`
            : doc.name;

          desc.textContent = doc.desc;

          section.appendChild(link);
          section.appendChild(desc);

          li.appendChild(section);

          elements.resultList.appendChild(li);
        }
      });

      //elements.numberResultsShowing.innerHTML = state.results.filtered.length;
    }
  }

  function displayResults() {
    if (!state.visible) {
      showResultsBox();
    }
    updateResultList();
  }

  function search() {
    if (elements.input.value.length > 0) {
      elements.searchLabel.classList.add("hidden");
    } else {
      elements.searchLabel.classList.remove("hidden");
    }

    if (elements.input.value.length >= MIN_SEARCH_LENGTH) {
      const initial = index.search(elements.input.value);

      if (initial.length === 0) {
        const wildcardResults = index.search(`${elements.input.value}*`);
        state.results.full =
          wildcardResults.length > 0
            ? wildcardResults
            : index.search(`*${elements.input.value}*`);
      } else {
        state.results.full = initial;
      }

      displayResults();
    }
  }
}

document.addEventListener("DOMContentLoaded", initializeSearch);