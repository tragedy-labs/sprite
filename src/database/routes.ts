export enum Routes {
  QUERY = 'query',
  COMMAND = 'command',
  BEGIN = 'begin',
  COMMIT = 'commit',
  ROLLBACK = 'rollback',
  EXISTS = 'exists'
}
