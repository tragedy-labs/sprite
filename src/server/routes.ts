export enum Routes {
  READY = 'ready',
  COMMAND = 'server',
  DATABASES = 'databases',
  EXISTS = 'exists'
}
