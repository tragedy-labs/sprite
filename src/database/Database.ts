import { Routes } from './routes.js';
import { EXPLAIN, SELECT_SCHEMA } from './commands/index.js';
import { Rest } from '../rest/Rest.js';
import {
  ArcadeTransactionIsolationLevel,
  SpriteTransaction
} from '../transaction/SpriteTransaction.js';
import { DatabaseSession } from '../session/DatabaseSession.js';
import { HeaderKeys } from '../rest/SpriteHeaders.js';
import { ArcadeSqlExplanation } from '@/types/database.js';

export enum Dialect {
  SQL = 'sql',
  SQLSCRIPT = 'sqlscript',
  GRAPHQL = 'graphql',
  CYPHER = 'cypher',
  GREMLIN = 'gremlin',
  MONGO = 'mongo'
}

export type ArcadeQueryParameters = Record<string, boolean | string | number>;

export type ArcadeCommand =
  | string
  | { command: string; params: Record<string, any> };

/**
 * The Query languages supported by ArcadeDB, supplied as a parameter
 * to `SpriteDatabase.query()` and `SpriteDatabase.command()`
 */
export type ArcadeSupportedQueryLanguages =
  | 'sql'
  | 'sqlscript'
  | 'graphql'
  | 'cypher'
  | 'gremlin'
  | 'mongo';

/**
 * Static methods for common database operations.
 */
class Database {
  /**
   * Static method to execute a command on the database.s
   * @param session The {@link DatabaseSession `DatabaseSession`} to execute the command on.
   * @param language The language of the command.
   * @param command The command to execute.
   * @param transaction The transaction to execute the command in.
   * @returns The result of the command.
   */
  public static command = async <T>(
    session: DatabaseSession,
    language: ArcadeSupportedQueryLanguages,
    command: ArcadeCommand,
    parameters?: Record<string, any>
  ): Promise<T> => {
    try {
      return await Rest.postJson(
        Routes.COMMAND,
        {
          language,
          command,
          params: parameters
        },
        session
      );
    } catch (error) {
      throw new Error(
        `Could not perform command on database: ${session.databaseName}`,
        { cause: error }
      );
    }
  };
  /**
   * Static method to query the database.
   * @param session The {@link DatabaseSession `DatabaseSession`} to execute the query on.
   * @param language The language of the query.
   * @param command The query to execute.
   * @param parameters
   * @returns The result-set of the query.
   */
  public static query = async <T>(
    session: DatabaseSession,
    language: ArcadeSupportedQueryLanguages,
    command: string,
    parameters?: ArcadeQueryParameters
  ): Promise<T[]> =>
    Rest.postJson<T[]>(
      Routes.QUERY,
      { language, command, params: parameters },
      session
    );
  /**
   * Static method to get the schema of the database.
   * @param session The {@link DatabaseSession `DatabaseSession`} to target for the schema.
   * @returns The schema of the database.
   */
  public static getSchema = async (session: DatabaseSession): Promise<any> => {
    try {
      return await this.query(session, Dialect.SQL, SELECT_SCHEMA);
    } catch (error) {
      throw new Error(
        `Could not retreive schema from database: ${session.databaseName}`,
        {
          cause: error
        }
      );
    }
  };
  /**
   * Static method to explain a query.
   * @param session The {@link DatabaseSession `DatabaseSession`} to target for the explanation.
   * @param sql The SQL query to explain.
   * @returns The explanation of the query.
   */
  public static explain = async (
    session: DatabaseSession,
    sql: string
  ): Promise<ArcadeSqlExplanation> => {
    try {
      const result = await this.query<ArcadeSqlExplanation>(session, Dialect.SQL, EXPLAIN(sql));
      return result[0];
    } catch (error) {
      throw new Error(`Could not retreive explanation for ${sql}.`, {
        cause: error
      });
    }
  };
  /**
   * Static method to execute a transaction.
   * @param session The {@link DatabaseSession `DatabaseSession`} to execute the transaction on.
   * @param callback The callback to execute within transaction context.
   * @param somethingElse Some other parameter to pass to the transaction.
   * @returns A touple containing result of the transaction, and a boolean indicating if the transaction was committed or rolled-back.
   */
  public static transaction = async <T = void>(
    session: DatabaseSession,
    callback: (trx: SpriteTransaction) => Promise<T>,
    isolationLevel?: ArcadeTransactionIsolationLevel
  ): Promise<[boolean, T]> => {
    const trx = await this.beginTransaction(session, isolationLevel);
    try {
      const result = await callback(trx);
      if (trx.rolledBack) {
        return [false, result];
      } else {
        await trx.commit();
        return [true, result];
      }
    } catch (error) {
      // Attempt to rollback the transaction
      try {
        await trx.rollback();
        throw new Error(`Transaction ${trx.id} failed and was rolled back.`, {
          cause: error
        });
      } catch (rollbackError) {
        throw new Error(
          `Transaction ${trx.id} failed and rollback also failed.`,
          { cause: rollbackError }
        );
      }
    }
  };
  /**
   * Static method to begin a transaction.
   * @param session The {@link DatabaseSession `DatabaseSession`} to begin the transaction on.
   * @param isolationLevel The isolation level of the transaction.
   * @returns A {@link SpriteTransaction `SpriteTransaction`} instance for the target session.
   */
  public static beginTransaction = async (
    session: DatabaseSession,
    isolationLevel?: ArcadeTransactionIsolationLevel
  ): Promise<SpriteTransaction> => {
    try {
      // 'READ_COMMITTED' is default in ARCADEDB,
      // so we don't bother sending that
      const response = await Rest.post(
        Routes.BEGIN,
        isolationLevel === 'REPEATABLE_READ'
          ? JSON.stringify({ isolationLevel })
          : null,
        session
      );

      if (response.status !== 204) {
        throw new Error(
          `Server returned an unexpected response. Status: ${response.status} / ${response.statusText}.`
        );
      }

      const transactionId = response.headers?.get(HeaderKeys.ArcadeSessionId);

      // because the headers could be null
      // TODO: this isn't a good way to check for the transactionId
      if (!transactionId || typeof transactionId !== 'string') {
        throw new Error('Invalid transaction key received from server.');
      } else {
        return new SpriteTransaction(session, transactionId);
      }
    } catch (error) {
      throw new Error(
        `Unable to begin transaction in database "${session.databaseName}".`,
        { cause: error }
      );
    }
  };
  /**
   * Static method to commit a transaction.
   * @param session The {@link DatabaseSession `DatabaseSession`} to commit the transaction on.
   * @param transaction The id of the transaction to commit.
   * @returns `true` if the transaction was committed.
   */
  public static commitTransaction = async (
    session: DatabaseSession,
    transaction: SpriteTransaction
  ) => {
    try {
      const result = await Rest.post(Routes.COMMIT, null, session, transaction);
      if (result.status === 204) {
        return true;
      } else {
        throw new Error(
          `Unexpected response from the server when attemping to commit transaction ${transaction.id}`
        );
      }
    } catch (error) {
      throw new Error(`Unable to commit transaction ${transaction.id}`, {
        cause: error
      });
    }
  };
  /**
   * Static method to rollback a transaction.
   * @param session The {@link DatabaseSession `DatabaseSession`} to rollback the transaction on.
   * @param transaction The id of the transaction to rollback.
   * @returns `true` if the transaction was rolled back.
   */
  public static rollbackTransaction = async (
    session: DatabaseSession,
    transaction: SpriteTransaction
  ) => {
    try {
      const result = await Rest.post(
        Routes.ROLLBACK,
        null,
        session,
        transaction
      );
      if (result.status === 204) {
        return true;
      } else {
        throw new Error(
          `Unexpected response from the server when attemping to rollback transaction ${transaction.id}`
        );
      }
    } catch (error) {
      throw new Error(`Unable to rollback transaction ${transaction.id}`, {
        cause: error
      });
    }
  };
  /**
   * Check to see if the database exists.
   * @param session The session to use to check for the database.
   * @param databaseName The name of the database to check for existence.
   * @returns `true` if database exists, `false` if not
   * @throws `Error` if the existence of the database could not be verified.
   */
  public static exists = async (
    session: DatabaseSession,
    databaseName: string
  ): Promise<boolean> => {
    try {
      const response = await Rest.get(Routes.EXISTS, session);
      switch (response.status) {
        case 200: {
          const { result } = await response.json();
          if (typeof result === 'boolean') {
            return result;
          } else {
            throw new Error(
              `Recieved an unexpected result from the server, expected boolean, recieved: [${result}] which has a type of: ${typeof result}.`
            );
          }
        }
        case 400:
          throw new Error(
            `No database name was passed to the server. Recieved: [${databaseName}].`
          );
        default:
          throw new Error(
            `Received an unexpected status code from the server when attempting to check if database "${databaseName}" exists.`
          );
      }
    } catch (error) {
      throw new Error(
        `Encountered an error when checking to see if database "${databaseName}" exists`,
        { cause: error }
      );
    }
  };
}

export { Database };
