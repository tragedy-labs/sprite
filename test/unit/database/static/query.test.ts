// Lib
import { Database, Dialect } from '@/database/Database.js';
import { Routes } from '@/database/routes.js';

// Testing
import {
  TestDatabaseSession as SESSION,
  headers,
  variables
} from '@test/variables.js';

const ENDPOINT = `${variables.address}${variables.apiRoute}${Routes.QUERY}/${variables.databaseName}`;

describe('Database.query()', () => {
  it(`should make a properly formatted POST request to ${ENDPOINT}`, async () => {
    jest.spyOn(global, 'fetch').mockResolvedValueOnce({
      status: 200,
      json: async () => variables.jsonResponse
    } as Response);
    await Database.query(SESSION, Dialect.GREMLIN, variables.nonEmptyString);

    const REQUEST_INIT: RequestInit = {
      method: 'POST',
      headers,
      body: JSON.stringify({
        language: Dialect.GREMLIN,
        command: variables.nonEmptyString
      }),
      keepalive: true
    };

    expect(global.fetch).toHaveBeenCalledWith(ENDPOINT, REQUEST_INIT);
  });

  it('should handle a 200 response by returning the data from the response', async () => {
    jest.spyOn(global, 'fetch').mockResolvedValueOnce({
      status: 200,
      json: async () => variables.jsonResponse
    } as Response);
    const result = await Database.query(
      SESSION,
      Dialect.GREMLIN,
      variables.nonEmptyString
    );
    expect(result).toMatchSnapshot();
  });

  it('should handle a 400 response by throwing an error', async () => {
    jest.spyOn(global, 'fetch').mockResolvedValueOnce({
      status: 400
    } as Response);
    await expect(
      Database.query(SESSION, Dialect.GREMLIN, variables.nonEmptyString)
    ).rejects.toMatchSnapshot();
  });

  it('should handle a 500 response by throwing an error', async () => {
    jest.spyOn(global, 'fetch').mockResolvedValueOnce({
      status: 500,
      json: async () => ({
        error: 'Internal error',
        detail: 'Database \u0027school\u0027 is not available',
        exception: 'com.arcadedb.exception.DatabaseOperationException'
      })
    } as Response);
    await expect(
      Database.query(SESSION, Dialect.GREMLIN, variables.nonEmptyString)
    ).rejects.toMatchSnapshot();
  });

  it('should handle an unexpected status code by throwing an error', async () => {
    jest.spyOn(global, 'fetch').mockResolvedValueOnce({
      status: 418,
      statusText: variables.nonEmptyString
    } as Response);
    await expect(
      Database.query(SESSION, Dialect.GREMLIN, variables.nonEmptyString)
    ).rejects.toMatchSnapshot();
  });
});
