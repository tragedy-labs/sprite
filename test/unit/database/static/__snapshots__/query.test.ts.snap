// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Database.query() should handle a 200 response by returning the data from the response 1`] = `"test"`;

exports[`Database.query() should handle a 400 response by throwing an error 1`] = `[TypeError: response.json is not a function]`;

exports[`Database.query() should handle a 500 response by throwing an error 1`] = `[ArcadeDatabaseError: Internal error. Database 'school' is not available.]`;

exports[`Database.query() should handle an unexpected status code by throwing an error 1`] = `[TypeError: response.json is not a function]`;
